### PR TITLE
[libc] rework mutex

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
This PR rework the `Mutex` implementation.

- Prepare for `timed` and `pshared` support.
- Make conditional variable  use only normal futex-world-sized mutex for its internal wait queue.
- Adjust all appearance of mutexes accordingly.

### CODE REVIEW

File: libc/src/__support/threads/linux/mutex.h
Comment (QuarticCat) [inline libc/src/__support/threads/linux/mutex.h:36]: This enum seems to be redundant now.

**Review by QuarticCat — APPROVED** [review]
LGTM

**Comment by QuarticCat** [comment]
I'm wondering if it's better to do CAS during spinning.

**Comment by SchrodingerZhu** [comment]
> I'm wondering if it's better to do CAS during spinning.

Implementation of CAS typically generates a write event regardless of comparison. CAS loop may lead to significantly more cache controller events.

**Comment by QuarticCat** [comment]
Fair enough.

**Review by michaelrj-google — CHANGES_REQUESTED** [review]
I'm fine with the idea of having a raw mutex that other mutexes inherit from, but I'm not sure if the implementation of `raw_mutex` here is correct. Here's the posix standard on the behavior of the lock, trylock, and unlock functions: https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_lock.html

My main concern is around `spin`. Generally it's better to let the kernel handle waiting (and possibly spinning) when possible, also 100 loops will take practically no time.

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (QuarticCat) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: Modern mutex implementations often do a few spins before waiting. Because waiting is slow. This approach can greatly improve performance under light contention IIRC.

Comment (michaelrj-google) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: that makes a lot of sense, do you have an example I could look at so I can understand this design better?

Comment (QuarticCat) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: I can only remember this [blog post](https://matklad.github.io/2020/01/04/mutexes-are-faster-than-spinlocks.html). But I do have seen this opinion in many places. Just think of many small critical regions contending for a single mutex. In this case, a few spins can avoid many syscalls.

Comment (QuarticCat) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: BTW, this PR looks pretty much similar to Rust's mutex implementation.

FYI: https://github.com/rust-lang/rust/blob/8387315ab3c26a57a1f53a90f188f0bc88514bca/library/std/src/sys/sync/mutex/futex.rs

File: libc/src/__support/threads/linux/mutex.h
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/mutex.h:28]: I think [gettid](https://man7.org/linux/man-pages/man2/gettid.2.html) is also using `pid_t` as its return value.

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: I left the `spin_count` as a configurable option. One can set `LIBC_CONF_RAW_MUTEX_DEFAULT_SPIN_COUNT` to `0` to disable such behavior.

Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: Notice that the spin loops continues only if the lock is hold by someone and contention is not heavy. In that case, the spinning thread knows that there is someone making progress in their critical section and the spinning one will likely obtain the lock once that critical section is over. Therefore, if the critical section is short, considerable syscalls can be avoided.

Notice that `CndVar`, `RwLock` exhibits waiting queues protected by mutexes. The critical section concerning the update of the queue is actually small. Hence, I think it may be a good choice to spin several rounds before going to sleep (futex_wait). 

I currently choose `100` as the default value because it aligns with the choice of Rust's standard library.

File: libc/src/__support/threads/linux/mutex.h
Comment (michaelrj-google) [reply libc/src/__support/threads/linux/mutex.h:28]: 👍

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (michaelrj-google) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: This does look very similar to the rust implementation, I have two comments about that: First, make sure that you're not just blindly copying what they do, it might not match the use case or semantics of libc. Second, make sure that you're not copying the code because that could have license implications. It should be fine as long as you're primarily writing the code yourself (but also I am not a lawyer).

As for leaving the spinning in, that's fine with me now that I understand the purpose of it.

Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: To my understanding, besides spin, Rust's mutex and rwlock implementation itself is largely similar to `bionic`'s implementation (simplified because Rust does not support complicated pthread features). See 
- https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/pthread_mutex.cpp. 
- https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/pthread_rwlock.cpp. 

The code here  is not directly copied from Rust, whose `RawMutex` does not have pshared and timeout support (In fact, I was told that Rust's standard library is not designed to support forking the process without replace the executable image).

However, I do think Rust has good clean and tidy codebase to study. Its platform abstraction layer can be very useful when learning about fundamental runtime operations.

Comment (michaelrj-google) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: taking design inspiration from rust should be fine, and same for bionic. This code doesn't seem like it will be a problem but I wanted to make sure you were aware.

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (michaelrj-google) [inline libc/src/__support/threads/linux/raw_mutex.h:103]: nit: The behavior here is a little hard to parse since you're relying on logical operand shortcutting to call `try_lock` and then if it fails call `lock_slow`. I would prefer it if you changed it to `if(try_lock()) return true; else return lock_slow;` or something similar (you might not need the else).

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (michaelrj-google) [inline libc/src/__support/threads/linux/raw_mutex.h:78]: You should either wrap `ETIMEDOUT` in parentheses or move the negative sign over to the other side. As a macro it could be defined in a way that evaluates to the correct result but doesn't handle a negative sign properly without parentheses (e.g. `#define ETIMEDOUT 85+1`)

File: libc/src/__support/threads/linux/mutex.h
Comment (michaelrj-google) [inline libc/src/__support/threads/linux/mutex.h:64]: if you have this set as the default parameter, you shouldn't also pass it here.

File: libc/test/src/__support/threads/linux/raw_mutex_test.cpp
Comment (michaelrj-google) [inline libc/test/src/__support/threads/linux/raw_mutex_test.cpp:1]: nit: needs license header

**Review by michaelrj-google — COMMENTED** [review]
some comments, but overall looks good. I would like to get a chance to review and approve one last time before this lands thought

**Comment by SchrodingerZhu** [comment]
> some comments, but overall looks good. I would like to get a chance to review and approve one last time before this lands thought

Sure!

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/raw_mutex.h]: fixed

File: libc/test/src/__support/threads/linux/raw_mutex_test.cpp
Comment (SchrodingerZhu) [reply libc/test/src/__support/threads/linux/raw_mutex_test.cpp:9]: fixed

File: libc/src/__support/threads/linux/mutex.h
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/mutex.h]: fixed

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/raw_mutex.h]: fixed

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (nickdesaulniers) [reply libc/src/__support/threads/linux/raw_mutex.h:51]: heh, I was just poking through the book Rust Atomics and Lock: Low-Level Concurrency in Practice by Mara Bos for inspiration, coincidentally.

File: libc/src/__support/threads/linux/mutex.h
Comment (nickdesaulniers) [inline libc/src/__support/threads/linux/mutex.h:54]: Should we check the return value?

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (nickdesaulniers) [inline libc/src/__support/threads/linux/raw_mutex.h:59]: I'd rather you use `__builtin_expect` at the call sites of `lock_slow` rather than the `cold` fn attr, which will elide many optimizations within this function.  For `lock_slow` and `wake`.

File: libc/test/src/__support/threads/linux/raw_mutex_test.cpp
Comment (nickdesaulniers) [inline libc/test/src/__support/threads/linux/raw_mutex_test.cpp:71]:
```suggestion
  // child early exit to avoid output pollution
```

**Review by nickdesaulniers — COMMENTED** [review]
Thanks for the patch! Nice work!

File: libc/test/src/__support/threads/linux/raw_mutex_test.cpp
Comment (SchrodingerZhu) [reply libc/test/src/__support/threads/linux/raw_mutex_test.cpp:46]: Oops, one `timeout` ought to be removed.
```c++
  ASSERT_FALSE(mutex.lock(*timeout));    // this lock is a deadlock, but it is backed by a timeout
  ASSERT_TRUE(mutex.unlock());                 // this unlocks the previous locked mutex            
  ASSERT_TRUE(mutex.lock());                      // this line (with timeout removed),
  ASSERT_TRUE(mutex.unlock());                 // and this line checks that the mutex still works fine 
  // If a lock can be acquired directly, expired timeout will not count.
  ASSERT_TRUE(mutex.lock(*timeout));     // this is an expired lock, but it should succeed
  ASSERT_TRUE(mutex.unlock());
  ```

File: libc/src/__support/threads/linux/CndVar.cpp
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/CndVar.cpp:32]: I think so, hh. `cpp::lock_guard` is more generic. Guess we can remove `MutexLock` later on, since they serve the same functionality and `MutexLock` is type specific.

Comment (nickdesaulniers) [reply libc/src/__support/threads/linux/CndVar.cpp:32]: oh, I forgot to request that `MutexLock` get removed in c4a3d184db5fdffe798208b8281dfe944616f9ed.

Comment (nickdesaulniers) [reply libc/src/__support/threads/linux/CndVar.cpp:32]: Mind splitting the removal of that and the fixup to libc/src/__support/threads/linux/CndVar.cpp of my mistake to a distinct PR from this one?

File: libc/src/__support/threads/linux/raw_mutex.h
Comment (nickdesaulniers) [inline libc/src/__support/threads/linux/raw_mutex.h:48]: This changes the signedness of `spin_count`. Do we want to use `unsigned` instead?

File: libc/src/__support/File/dir.h
Comment (michaelrj-google) [inline libc/src/__support/File/dir.h:72]: all of the functions in this header should be marked `LIBC_INLINE`

**Review by michaelrj-google — COMMENTED** [review]
Overall LGTM though wait for Nick's approval

**Review by michaelrj-google — APPROVED** [review]
Overall LGTM though wait for Nick's approval

File: libc/src/__support/threads/linux/CndVar.cpp
Comment (SchrodingerZhu) [reply libc/src/__support/threads/linux/CndVar.cpp:32]: ok

File: libc/src/__support/File/dir.h
Comment (SchrodingerZhu) [reply libc/src/__support/File/dir.h:72]: Just to confirm: including declarations with separate definitions in another TU?

**Review by nickdesaulniers — APPROVED** [review]

**Pull request merged**
